### PR TITLE
Improve callback documentation for `outputs` and `accumulate_grad_batches` (Resolves #15315)

### LINKED ISSUE #15315: Misleading outputs (loss) in callback when using `accumulate_grad_batches` > 1 in trainer
## 📚 Documentation

Some of the callback entry points/hooks (e.g. `on_train_batch_end`) have an `outputs` argument that contains the loss returned from the `training_step` and `validation_step`. This value does not match the loss in the trainer when `accumulate_grad_batches` > 1. I spoke with @rohitgr7 about this on slack and he said that this is expected behavior and that the loss is normalized internally. We agreed the best fix was a documentation PR alerting users of this possibility, and creating an issue so people could find it if they also experience it. I will submit a PR updating the documentation for the callback entry points explaining this behavior.

### PR DESCRIPTION
**Pull request by rbracco**
## What does this PR do?

Adds brief documentation explaining the `outputs` parameter to callback hooks, and that potentially unexpected behavior occurs when using `accumulate_grad_batches` > 1 

Fixes #15315 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rbracco** [comment]
I'm not sure of the best place to put the documentation update, so I chose what I thought was the best spot but I'm happy to move it elsewhere. I built docs locally and I think it looks reasonable but let me know if changes are needed 

### Before
image

### After
image

File: docs/source-pytorch/extensions/callbacks.rst [hunk 1]
```diff
@@ -138,6 +138,11 @@ Callback API
 ************
 Here is the full API of methods available in the Callback base class.
 
+.. note::
+
+    Outputs: Hooks that occur on batch end (e.g. ``on_train_batch_end``) will receive any values returned from that step (such as the loss) in the form of a dictionary named ``outputs``.
+    In the case ``accumulate_grad_batches`` > 1, these values will be normalized by Lightning so that the value provided in ``outputs`` will be the value returned from the training/validation/predict step divided by ``accumulate_grad_batches``
```
Comment (rohitgr7) [inline docs/source-pytorch/extensions/callbacks.rst:144]: 1. only loss is scaled... Other outputs are kept as it is.
2. Also it's the case only with `training_step`. For rest it is same.

I'd recommend adding it to the `training_step` docstring too.

Comment (rbracco) [reply docs/source-pytorch/extensions/callbacks.rst:144]: > 1. only loss is scaled... Other outputs are kept as it is.
> 2. Also it's the case only with `training_step`. For rest it is same.
> 
> I'd recommend adding it to the `training_step` docstring too.

Thanks, I'll fix and submit. By `training_step` do you mean the `on_train_batch_end` hook? Or do you want me adjust the actual `training_step` docstring?

Comment (rohitgr7) [reply docs/source-pytorch/extensions/callbacks.rst:144]: adjust actual training_step hook. just add a note that the loss returned from here will be normalized internally.

Comment (rbracco) [reply docs/source-pytorch/extensions/callbacks.rst]: Updated. I tried to find the best location in the training step docstring, and I updated the callback docs to include the specific details you mentioned. Please let me know if further fixes are needed and thank you for the time and help.

File: docs/source-pytorch/extensions/callbacks.rst [hunk 2]
```diff
@@ -141,7 +141,7 @@ Here is the full API of methods available in the Callback base class.
 .. note::
 
     Outputs: Hooks that occur on batch end (e.g. ``on_train_batch_end``) will receive any values returned from that step (such as the loss) in the form of a dictionary named ``outputs``.
-    In the case ``accumulate_grad_batches`` > 1, these values will be normalized by Lightning so that the value provided in ``outputs`` will be the value returned from the training/validation/predict step divided by ``accumulate_grad_batches``
+    For ``on_train_batch_end``, in the case ``accumulate_grad_batches`` > 1, ``outputs["loss"]`` is normalized by Lightning and will be the same as the value returned from the training/validation/predict step divided by ``accumulate_grad_batches``.
```
Comment (rohitgr7) [inline docs/source-pytorch/extensions/callbacks.rst:144]:
```suggestion
    ``outputs["loss"]`` here will be the normalized value w.r.t ``accumulate_grad_batches`` of the loss returned from ``training_step``.
```

File: src/pytorch_lightning/core/module.py [hunk 3]
```diff
@@ -661,6 +662,10 @@ def training_step(self, batch, batch_idx, hiddens):
         Note:
             The loss value shown in the progress bar is smoothed (averaged) over the last values,
             so it differs from the actual loss returned in train/validation step.
+
+        Note:
+            When ``accumulate_grad_batches`` > 1, the loss returned from the training step will be automatically
+            normalized internally by dividing by ``accumulate_grad_batches``.
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/module.py:668]:
```suggestion
            When ``accumulate_grad_batches`` > 1, the loss returned here will be automatically
            normalized by ``accumulate_grad_batches`` internally.
```

File: src/pytorch_lightning/core/module.py [hunk 4]
```diff
@@ -615,14 +615,15 @@ def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:  # type: igno
...
-            - :class:`~torch.Tensor` - The loss tensor
+            - :class:`~torch.Tensor` - The loss tensor.
             - ``dict`` - A dictionary. Can include any keys, but must include the key ``'loss'``
             - ``None`` - Training will skip to the next batch. This is only for automatic optimization.
                 This is not supported for multi-GPU, TPU, IPU, or DeepSpeed.
 
         In this step you'd normally do the forward pass and calculate the loss for a batch.
         You can also do fancier things like multiple forward passes or something model specific.
 
+
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/module.py:626]:
```suggestion
```

File: src/pytorch_lightning/core/module.py [hunk 5]
```diff
@@ -615,14 +615,15 @@ def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:  # type: igno
         Return:
             Any of.
 
-            - :class:`~torch.Tensor` - The loss tensor
+            - :class:`~torch.Tensor` - The loss tensor.
```
Comment (rohitgr7) [inline src/pytorch_lightning/core/module.py:618]:
```suggestion
            - :class:`~torch.Tensor` - The loss tensor
```

**Review by rohitgr7 — APPROVED** [review]

File: docs/source-pytorch/extensions/callbacks.rst [hunk 6]
```diff
@@ -138,6 +138,11 @@ Callback API
 ************
 Here is the full API of methods available in the Callback base class.
 
+.. note::
+
+    Outputs: Hooks that occur on batch end (e.g. ``on_train_batch_end``) will receive any values returned from that step (such as the loss) in the form of a dictionary named ``outputs``.
+    ``outputs["loss"]`` here will be the normalized value w.r.t ``accumulate_grad_batches`` of the loss returned from ``training_step``.
```
Comment (awaelchli) [inline docs/source-pytorch/extensions/callbacks.rst:144]: This is not the right place to put it IMO. 

1. This note is inserted before we explain the Callback: "The Callback class is the base for all the callbacks in Lightning". This seems distracting.
2. I propose to add a simple sentence explaining this in the individual hook documentation. When a user implements the  on_train_batch_end, they would go to the documentation of that hook if they have any question about its signature.

Comment (rbracco) [reply docs/source-pytorch/extensions/callbacks.rst]: @awaelchli, thanks for the suggestion, this change has been pushed and is ready for review.

File: src/pytorch_lightning/callbacks/callback.py [hunk 7]
```diff
@@ -82,7 +82,12 @@ def on_train_batch_start(
     def on_train_batch_end(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", outputs: STEP_OUTPUT, batch: Any, batch_idx: int
     ) -> None:
-        """Called when the train batch ends."""
+        """Called when the train batch ends.
+
+        Note:
+            The value ``outputs["loss"]`` here will be the normalized value w.r.t ``accumulate_grad_batches`` of the
+            loss returned from ``training_step``.
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/callback.py:89]: @rbracco I pushed your note to the docstring here directly. It will show up automatically in the docs where you originally put it.

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
